File: sites/registry.py
from typing import Dict, Type, Optional
from .base import BaseSiteAdapter
from .madara import ManhuaPlusAdapter, ManhwaTopAdapter, MadaraAdapter, ManhuascanAdapter, ManhuaPlusTopAdapter
from .asura import AsuraAdapter
from .arena import ArenaAdapter
from .mgeko import MgekoAdapter


# Registry of all available site adapters
SITE_REGISTRY: Dict[str, Type[BaseSiteAdapter]] = {
    "arenascans.com": ArenaAdapter,
    "asurascanz.com": AsuraAdapter,
    "manhuaplus.com": ManhuaPlusAdapter,
    "manhwatop.com": ManhwaTopAdapter,
    "manhuascan.us": ManhuascanAdapter,
    "manhuaplus.top": ManhuaPlusTopAdapter,
    "mgeko.cc": MgekoAdapter,
}

# Site display names for UI
SITE_NAMES = {
    "arenascans.com": "Arenascans",
    "asurascanz.com": "Asura Scans",
    "manhuaplus.com": "ManhuaPlus",
    "manhwatop.com": "ManhwaTop",
    "manhuascan.us": "Manhuascan",
    "manhuaplus.top": "ManhuaPlus V2",
    "mgeko.cc": "Mgeko",
}


def get_adapter(domain: str) -> Optional[BaseSiteAdapter]:
    """Get a site adapter by domain"""
    adapter_class = SITE_REGISTRY.get(domain)
    if adapter_class:
        return adapter_class()
    return None


def get_all_sites() -> Dict[str, str]:
    """Get all available sites as {domain: name}"""
    return {domain: SITE_NAMES.get(domain, domain) for domain in SITE_REGISTRY}


def is_supported_site(url: str) -> bool:
    """Check if a URL is from a supported site"""
    from urllib.parse import urlparse
    parsed = urlparse(url)
    domain = parsed.netloc.replace('www.', '')
    return domain in SITE_REGISTRY


def detect_site(url: str) -> Optional[str]:
    """Detect which site a URL is from"""
    from urllib.parse import urlparse
    parsed = urlparse(url)
    domain = parsed.netloc.replace('www.', '')
    
    # Check exact match
    if domain in SITE_REGISTRY:
        return domain
    
    # Check if any registered domain is contained in the URL domain
    for registered_domain in SITE_REGISTRY:
        if registered_domain in domain:
            return registered_domain
    
    return None

File: manga-reader-rn/src/lib/scraper/registry.ts
import { BaseSiteAdapter } from "./base";
import { MadaraAdapter, ManhuaPlusAdapter, ManhwaTopAdapter, ManhuascanAdapter, ManhuaPlusTopAdapter } from "./madara";
import { AsuraAdapter } from "./asura";
import { ArenaAdapter } from "./arena";
import { MgekoAdapter } from "./mgeko";
import { ThunderscansAdapter, RoliascanAdapter } from "./thunderscans";

const SITE_REGISTRY: Record<string, () => BaseSiteAdapter> = {
  "arenascans.com": () => new ArenaAdapter(),
  "asurascanz.com": () => new AsuraAdapter(),
  "manhuaplus.com": () => new ManhuaPlusAdapter(),
  "manhuascan.us": () => new ManhuascanAdapter(),
  "mgeko.cc": () => new MgekoAdapter(),
  "en-thunderscans.com": () => new ThunderscansAdapter(),
  "roliascan.com": () => new RoliascanAdapter(),
  "manhwatop.com": () => new ManhwaTopAdapter(),
  "manhuaplus.top": () => new ManhuaPlusTopAdapter(),
};

const SITE_NAMES: Record<string, string> = {
  "arenascans.com": "Arenascans",
  "asurascanz.com": "Asura Scans",
  "manhuaplus.com": "ManhuaPlus",
  "manhuascan.us": "Manhuascan",
  "mgeko.cc": "Mgeko",
  "en-thunderscans.com": "Thunder Scans",
  "roliascan.com": "Roliascan",
  "manhwatop.com": "ManhwaTop",
  "manhuaplus.top": "ManhuaPlusTop",
};

export function getAdapter(domain: string): BaseSiteAdapter | null {
  const factory = SITE_REGISTRY[domain];
  return factory ? factory() : null;
}

export function getAllSites(): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [domain, name] of Object.entries(SITE_NAMES)) {
    result[domain] = name;
  }
  return result;
}

export function isSupportedSite(url: string): boolean {
  return !!detectSite(url);
}

export function detectSite(url: string): { domain: string; name: string } | null {
  for (const [domain, name] of Object.entries(SITE_NAMES)) {
    if (url.includes(domain)) {
      return { domain, name };
    }
  }
  return null;
}
